### PR TITLE
Add trimming in wgbs workflows

### PR DESCRIPTION
**Pull request by AnzeLovse**
This PR adds Trimming to WGBS workflow. It is recommended to apply clipping of Illumina adapters and specific tails introduced in some library preparation kits for bisulfite sequencing. Prior tests show an increase from 30% mapped reads to more than 70% mapped reads which makes it a useful addition to the workflow.

## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/workflows/wgbs.yml [hunk 1]
```diff
@@ -47,6 +48,64 @@
...
+    - name: trimming
+      label: Parameters for the trimming step
+      group:
+        - name: adapters
+          label: Adapter sequences
+          type: data:seq:nucleotide
+          required: false
+          description: |
+            Adapter sequence in FASTA format that will be removed from the read.
+            This field as well as 'Seed mismatches' and 'Simple clip threshold' parameters are needed to perform Illuminacliping.
```
Comment (romunov) [inline resolwe_bio/processes/workflows/wgbs.yml]: Is this designation that the parameter is needed for illuminaclipping necessary? If we need this information, would having another group of parameters help?

If yes, consider clipping line width to something like 72 characters.

File: resolwe_bio/processes/workflows/wgbs.yml [hunk 2]
```diff
@@ -364,6 +439,108 @@
...
         otherwise need to be decoded. Default is STRICT.
+    - name: trimming
+      label: Parameters for the trimming step
+      group:
+        - name: adapters
+          label: Adapter sequences
+          type: data:seq:nucleotide
+          required: false
+          description: |
+            Adapter sequence in FASTA format that will be removed from the read.
```
Comment (romunov) [inline resolwe_bio/processes/workflows/wgbs.yml]: Same comment about line length as above.

File: resolwe_bio/processes/workflows/wgbs.yml [hunk 3]
```diff
@@ -364,6 +439,108 @@
...
+        - name: simple_clip_threshold
+          label: Simple clip threshold
+          type: basic:integer
+          disabled: >
+            !trimming.adapters
+          required: false
+          description: |
+            Specifies how accurate the match between any adapter etc. sequence must be against a read.
+            This field as well as 'Adapter sequences' and 'Seed mismatches' parameter are needed to perform Illuminacliping.
+        - name: min_adapter_length
```
Comment (romunov) [inline resolwe_bio/processes/workflows/wgbs.yml]: I'm just trying to understand the reasoning behind different parameters for trimming for paired and single-end workflows.

File: resolwe_bio/processes/workflows/wgbs.yml (hunk 1 above)
Comment (AnzeLovse) [reply resolwe_bio/processes/workflows/wgbs.yml]: It is necessary to explain that all of those fields are required in order to perform illuminaclipping as those are all set as optional. Separating the illuminaclip parameters is more in line with the original process. Do you think that in this case it is sufficient to explain which parameters are needed only in the description for adapters input?

File: resolwe_bio/processes/workflows/wgbs.yml (hunk 3 above)
Comment (AnzeLovse) [reply resolwe_bio/processes/workflows/wgbs.yml]: Each uses different trimmomatic process, one is intended for single-end reads and the other for paired-end reads and some of the parameters are not used in the first one.

File: resolwe_bio/processes/workflows/wgbs.yml [hunk 4]
```diff
@@ -47,6 +48,75 @@
         performance when processing a BAM file in which
         variable-length data (read, qualities, tags) do not
         otherwise need to be decoded. Default is STRICT.
+    - name: illuminaclip
+      label: Illuminaclip
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/wgbs.yml]: The label should `Adapter trimming`. `Illuminaclip` is a subcommand name of Trimmomatic, and it doesn't say much to the layperson.

Comment (AnzeLovse) [reply resolwe_bio/processes/workflows/wgbs.yml]: Fixed. I have also modified the description to match this label.

**Pull request merged**